### PR TITLE
Better package names for Cucumber Junit reports

### PR DESCRIPTION
**Pull request by mikesalvia**
## Proposed changes

[//]: # (Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. If it fixes a bug or resolves a feature request, be sure to link to that issue.)

Create consistent package names in Junit reports created from Cucumber tests, this will be especially helpful as I am developing a method to combine re-run and primary run Junit reports for accurate # reporting in Jenkins.

Screen Shot 2020-06-24 at 12 05 53 PM

Above is how the device/version portions of the package name look in iOS using Browserstack.

Screen Shot 2020-06-24 at 12 06 15 PM

Above is how the device/version portions of the package name look in Android using Browserstack.

I consider this a bug fix but could be an improvement depending on how it is looked at.

In Android, it appears the `deviceUUID` is used which is not consistent, this code uses other items from `runner.capabilities` to construct a package name that will be as readable as the one for iOS.

## Types of changes

[//]: # (What types of changes does your code introduce to WebdriverIO?)
[//]: # (_Put an `x` in the boxes that apply_)

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## Checklist

[//]: # (_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._)

- [x] I have read the [CONTRIBUTING](https://github.com/webdriverio/webdriverio/blob/master/CONTRIBUTING.md) doc
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)
- [x] I have added proper type definitions for new commands (if appropriate)

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

### CODE REVIEW

**Comment by mikesalvia** [comment]
@christian-bromann I found a couple of issues while testing cleaning it up now.

File: packages/wdio-junit-reporter/src/index.js [hunk 1]
```diff
@@ -158,8 +158,14 @@ class JunitReporter extends WDIOReporter {
     buildJunitXml (runner) {
         let builder = junit.newBuilder()
 
-        this.packageName = this.options.packageName ? `${runner.sanitizedCapabilities}-${this.options.packageName}` : runner.sanitizedCapabilities
-
+        if (runner.config.hostname.indexOf('browserstack') && runner.capabilities.os === 'Android') {
+          // NOTE: deviceUUID is used resulting in a everchanging package name in runner.sanitizedCapabilities when running Android tests under Browserstack. (i.e. ht79v1a03938.android.9)
+          let browserstackSanitizedCapabilities = runner.capabilities.device.replace(' ', '') + '.' + runner.capabilties.os + '.' + runner.capabilities.os_version.replace(' ', '_');
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.js:163]: better:
```suggestion
          const browserstackSanitizedCapabilities = (runner.capabilities.device + '.' + runner.capabilties.os + '.' + runner.capabilities.os_version).replace(' ', '');
```

File: packages/wdio-junit-reporter/src/index.js [hunk 2]
```diff
@@ -160,7 +160,7 @@ class JunitReporter extends WDIOReporter {
 
         if (runner.config.hostname.indexOf('browserstack') && runner.capabilities.os === 'Android') {
           // NOTE: deviceUUID is used resulting in a everchanging package name in runner.sanitizedCapabilities when running Android tests under Browserstack. (i.e. ht79v1a03938.android.9)
-          let browserstackSanitizedCapabilities = runner.capabilities.device.replace(' ', '') + '.' + runner.capabilties.os + '.' + runner.capabilities.os_version.replace(' ', '_');
+          const browserstackSanitizedCapabilities = (runner.capabilities.device + '.' + runner.capabilties.os + '.' + runner.capabilities.os_version).replace(' ', '');
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.js:163]: Are you sure `runner.capabilities.device`, `runner.capabilties.os` and `runner.capabilities.os_version` are always present?

**Comment by mikesalvia** [comment]
@christian-bromann Changed also added a change for switching `.` to `_` in `os_version`.

**Comment by mikesalvia** [comment]
> Thanks for proposing the PR. Can you add a unit test for this?

Will do sorry about that.

**Comment by mikesalvia** [comment]
```
  162:1    error  Expected indentation of 12 spaces but found 10  indent
  163:1    error  Expected indentation of 12 spaces but found 10  indent
  163:212  error  A space is required after ','                   comma-spacing
  163:216  error  Extra semicolon                                 semi
  164:1    error  Expected indentation of 12 spaces but found 10  indent
  164:159  error  Extra semicolon                                 semi
  166:1    error  Expected indentation of 12 spaces but found 10  indent
  166:149  error  Extra semicolon                                 semi
  168:1    error  Trailing spaces not allowed                     no-trailing-
```
Got to clean this up too.

**Comment by christian-bromann** [comment]
If you need help adding the tests, let me know 😉

**Comment by mikesalvia** [comment]
@christian-bromann Can you point me in the right direction? I don't see any tests that confirm the package name currently, so my thought of copy and modify may not work.

**Comment by mikesalvia** [comment]
Good news though and for those that are following...This creates a much better Jenkins experience.

Screen Shot 2020-06-25 at 9 17 22 AM
Screen Shot 2020-06-25 at 9 18 10 AM

**Comment by mikesalvia** [comment]
@christian-bromann A side note, I have both a rerun service as well as a Junit re-run merge script that I think would fit somewhere in WDIO, should I make the rerun-service an external package item for the time being?

**Comment by christian-bromann** [comment]
> should I make the rerun-service an external package item for the time being?

Yeah .. bootstrapping all packages already consumes a lot of CPU and we currently are very careful adding any new packages to the monorepo.

**Comment by mikesalvia** [comment]
> Yeah .. bootstrapping all packages already consumes a lot of CPU and we currently are very careful adding any new packages to the monorepo.

Cool! Will do can you give some pointers regarding where to add the unit tests?

**Comment by christian-bromann** [comment]
@mikesalvia the unit tests should be added somewhere here: https://github.com/webdriverio/webdriverio/blob/master/packages/wdio-junit-reporter/tests/reporter.test.js

**Comment by mikesalvia** [comment]
@christian-bromann Sorry for the delay on the unit tests I found also that iOS has varying version #'s cause they don't guarantee minor version match in Browserstack.

File: packages/wdio-junit-reporter/src/index.js [hunk 3]
```diff
@@ -158,7 +158,15 @@ class JunitReporter extends WDIOReporter {
     buildJunitXml (runner) {
         let builder = junit.newBuilder()
 
-        this.packageName = this.options.packageName ? `${runner.sanitizedCapabilities}-${this.options.packageName}` : runner.sanitizedCapabilities
+        if (runner.config.hostname.indexOf('browserstack') > -1) {
+            console.log(JSON.stringify(runner.capabilities));
```
Comment (christian-bromann) [inline packages/wdio-junit-reporter/src/index.js:162]: fixing semicolon
```suggestion
            console.log(JSON.stringify(runner.capabilities))
```

**Comment by christian-bromann** [comment]
@mikesalvia please let me know if I can help with this? Don't hesitate to book a slot via our [WebdriverIO Office Hours](https://webdriver.io/blog/2020/07/01/office-hours.html) and we can sync on the unit tests.

**Comment by mikesalvia** [comment]
@christian-bromann It looks like these unit tests should satisfy the requirements. I didn't see the graphic get generated.

**Comment by mikesalvia** [comment]
Looks like we have a passing and unit tests maintain coverage! :-)

**Comment by christian-bromann** [comment]
Thanks for the contribution!

**Pull request merged**